gcs/constraints/smart_table.hh

```diff
@@ -0,0 +1,26 @@
+#ifndef GLASGOW_CONSTRAINT_SOLVER_GUARD_GCS_SMART_TABLE_HH
+#define GLASGOW_CONSTRAINT_SOLVER_GUARD_GCS_SMART_TABLE_HH
+
+#include <gcs/constraint.hh>
+#include <gcs/extensional.hh>
+#include <gcs/smart_entry.hh>
+#include <gcs/variable_id.hh>
+namespace gcs
+{
+    class SmartTable : public Constraint
+    {
+    private:
+        const std::vector<IntegerVariableID> & _vars;
```

Comment:
With the new clone constraints interface, these should be stored as values, not references. In fact, I think probably they should be a `shared_ptr` to the const data, and the user API should accept both a value and a  `shared_ptr`.